File: internal/storage/ingest.go
package storage

import (
	"bufio"
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// IngestAll reads all JSONL files from dataPath and ingests into db.
func IngestAll(db *sql.DB, dataPath string) error {
	for _, signal := range []string{"traces", "metrics", "logs"} {
		file := filepath.Join(dataPath, signal, signal+".jsonl")
		if _, err := os.Stat(file); os.IsNotExist(err) {
			continue
		}
		switch signal {
		case "traces":
			if err := ingestTraces(db, file); err != nil {
				return fmt.Errorf("ingesting traces: %w", err)
			}
		case "metrics":
			if err := ingestMetrics(db, file); err != nil {
				return fmt.Errorf("ingesting metrics: %w", err)
			}
		case "logs":
			if err := ingestLogs(db, file); err != nil {
				return fmt.Errorf("ingesting logs: %w", err)
			}
		}
	}
	return nil
}

func ingestTraces(db *sql.DB, file string) error {
	f, err := os.Open(file)
	if err != nil {
		return err
	}
	defer f.Close()

	tx, err := db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	stmt, err := tx.Prepare(`INSERT INTO traces (trace_id, span_id, parent_span_id, name, kind, start_time, end_time, duration_ns, status_code, service_name, attributes, date) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`)
	if err != nil {
		return err
	}
	defer stmt.Close()

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 1024*1024), 10*1024*1024)

	for scanner.Scan() {
		var batch struct {
			ResourceSpans []struct {
				Resource struct {
					Attributes []otlpAttr `json:"attributes"`
				} `json:"resource"`
				ScopeSpans []struct {
					Spans []struct {
						TraceID      string     `json:"traceId"`
						SpanID       string     `json:"spanId"`
						ParentSpanID string     `json:"parentSpanId"`
						Name         string     `json:"name"`
						Kind         int        `json:"kind"`
						StartTime    otlpNano   `json:"startTimeUnixNano"`
						EndTime      otlpNano   `json:"endTimeUnixNano"`
						Status       struct {
							Code int `json:"code"`
						} `json:"status"`
						Attributes []otlpAttr `json:"attributes"`
					} `json:"spans"`
				} `json:"scopeSpans"`
			} `json:"resourceSpans"`
		}

		if err := json.Unmarshal(scanner.Bytes(), &batch); err != nil {
			continue // skip malformed lines
		}

		for _, rs := range batch.ResourceSpans {
			svcName := extractServiceName(rs.Resource.Attributes)
			for _, ss := range rs.ScopeSpans {
				for _, span := range ss.Spans {
					startTime := span.StartTime.Time()
					endTime := span.EndTime.Time()
					durationNs := int64(0)
					if !startTime.IsZero() && !endTime.IsZero() {
						durationNs = endTime.Sub(startTime).Nanoseconds()
					}
					attrs, _ := json.Marshal(flattenAttrs(span.Attributes))

					_, err := stmt.Exec(
						span.TraceID, span.SpanID, nullStr(span.ParentSpanID),
						span.Name, span.Kind,
						startTime, endTime, durationNs,
						span.Status.Code, svcName,
						string(attrs), startTime.Format("2006-01-02"),
					)
					if err != nil {
						return fmt.Errorf("inserting span: %w", err)
					}
				}
			}
		}
	}
	return tx.Commit()
}

func ingestMetrics(db *sql.DB, file string) error {
	f, err := os.Open(file)
	if err != nil {
		return err
	}
	defer f.Close()

	tx, err := db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	stmt, err := tx.Prepare(`INSERT INTO metrics (metric_name, metric_type, value, timestamp, service_name, aggregation_temporality, is_monotonic, unit, attributes, date) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`)
	if err != nil {
		return err
	}
	defer stmt.Close()

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 1024*1024), 10*1024*1024)

	for scanner.Scan() {
		var batch struct {
			ResourceMetrics []struct {
				Resource struct {
					Attributes []otlpAttr `json:"attributes"`
				} `json:"resource"`
				ScopeMetrics []struct {
					Metrics []otlpMetric `json:"metrics"`
				} `json:"scopeMetrics"`
			} `json:"resourceMetrics"`
		}

		if err := json.Unmarshal(scanner.Bytes(), &batch); err != nil {
			continue
		}

		for _, rm := range batch.ResourceMetrics {
			svcName := extractServiceName(rm.Resource.Attributes)
			for _, sm := range rm.ScopeMetrics {
				for _, m := range sm.Metrics {
					for _, dp := range extractDataPoints(m) {
						attrs, _ := json.Marshal(flattenAttrs(dp.attributes))
						_, err := stmt.Exec(
							m.Name, dp.metricType, dp.value,
							dp.timestamp, svcName,
							dp.temporality, dp.monotonic,
							m.Unit, string(attrs),
							dp.timestamp.Format("2006-01-02"),
						)
						if err != nil {
							return fmt.Errorf("inserting metric: %w", err)
						}
					}
				}
			}
		}
	}
	return tx.Commit()
}

func ingestLogs(db *sql.DB, file string) error {
	f, err := os.Open(file)
	if err != nil {
		return err
	}
	defer f.Close()

	tx, err := db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	stmt, err := tx.Prepare(`INSERT INTO logs (timestamp, severity, severity_number, body, service_name, trace_id, span_id, attributes, date) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`)
	if err != nil {
		return err
	}
	defer stmt.Close()

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 1024*1024), 10*1024*1024)

	for scanner.Scan() {
		var batch struct {
			ResourceLogs []struct {
				Resource struct {
					Attributes []otlpAttr `json:"attributes"`
				} `json:"resource"`
				ScopeLogs []struct {
					LogRecords []struct {
						TimeUnixNano   otlpNano   `json:"timeUnixNano"`
						SeverityText   string     `json:"severityText"`
						SeverityNumber int        `json:"severityNumber"`
						Body           otlpValue  `json:"body"`
						TraceID        string     `json:"traceId"`
						SpanID         string     `json:"spanId"`
						Attributes     []otlpAttr `json:"attributes"`
					} `json:"logRecords"`
				} `json:"scopeLogs"`
			} `json:"resourceLogs"`
		}

		if err := json.Unmarshal(scanner.Bytes(), &batch); err != nil {
			continue
		}

		for _, rl := range batch.ResourceLogs {
			svcName := extractServiceName(rl.Resource.Attributes)
			for _, sl := range rl.ScopeLogs {
				for _, lr := range sl.LogRecords {
					ts := lr.TimeUnixNano.Time()
					attrs, _ := json.Marshal(flattenAttrs(lr.Attributes))
					_, err := stmt.Exec(
						ts, lr.SeverityText, lr.SeverityNumber,
						lr.Body.String(), svcName,
						nullStr(lr.TraceID), nullStr(lr.SpanID),
						string(attrs), ts.Format("2006-01-02"),
					)
					if err != nil {
						return fmt.Errorf("inserting log: %w", err)
					}
				}
			}
		}
	}
	return tx.Commit()
}

// otlpAttr represents an OTLP key-value attribute.
type otlpAttr struct {
	Key   string    `json:"key"`
	Value otlpValue `json:"value"`
}

// otlpValue represents an OTLP typed value.
type otlpValue struct {
	StringValue *string `json:"stringValue,omitempty"`
	IntValue    *string `json:"intValue,omitempty"`
	BoolValue   *bool   `json:"boolValue,omitempty"`
	DoubleValue *float64 `json:"doubleValue,omitempty"`
}

func (v otlpValue) String() string {
	if v.StringValue != nil {
		return *v.StringValue
	}
	if v.IntValue != nil {
		return *v.IntValue
	}
	if v.BoolValue != nil {
		if *v.BoolValue {
			return "true"
		}
		return "false"
	}
	if v.DoubleValue != nil {
		return fmt.Sprintf("%g", *v.DoubleValue)
	}
	return ""
}

// otlpNano handles nanosecond timestamps that may be strings or integers.
type otlpNano int64

func (n *otlpNano) UnmarshalJSON(b []byte) error {
	var s string
	if err := json.Unmarshal(b, &s); err == nil {
		var v int64
		fmt.Sscanf(s, "%d", &v)
		*n = otlpNano(v)
		return nil
	}
	var v int64
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	*n = otlpNano(v)
	return nil
}

func (n otlpNano) Time() time.Time {
	if n == 0 {
		return time.Time{}
	}
	return time.Unix(0, int64(n)).UTC()
}

type otlpMetric struct {
	Name        string          `json:"name"`
	Description string          `json:"description"`
	Unit        string          `json:"unit"`
	Sum         *otlpSum        `json:"sum,omitempty"`
	Gauge       *otlpGauge      `json:"gauge,omitempty"`
	Histogram   *otlpHistogram  `json:"histogram,omitempty"`
}

type otlpSum struct {
	DataPoints             []otlpDataPoint `json:"dataPoints"`
	AggregationTemporality int             `json:"aggregationTemporality"`
	IsMonotonic            bool            `json:"isMonotonic"`
}

type otlpGauge struct {
	DataPoints []otlpDataPoint `json:"dataPoints"`
}

type otlpHistogram struct {
	DataPoints             []otlpHistogramDP `json:"dataPoints"`
	AggregationTemporality int               `json:"aggregationTemporality"`
}

type otlpDataPoint struct {
	Attributes   []otlpAttr `json:"attributes"`
	TimeUnixNano otlpNano   `json:"timeUnixNano"`
	AsInt        *string    `json:"asInt,omitempty"`
	AsDouble     *float64   `json:"asDouble,omitempty"`
}

func (dp otlpDataPoint) Value() float64 {
	if dp.AsDouble != nil {
		return *dp.AsDouble
	}
	if dp.AsInt != nil {
		var v float64
		fmt.Sscanf(*dp.AsInt, "%f", &v)
		return v
	}
	return 0
}

type otlpHistogramDP struct {
	Attributes   []otlpAttr `json:"attributes"`
	TimeUnixNano otlpNano   `json:"timeUnixNano"`
	Count        *string    `json:"count,omitempty"`
	Sum          *float64   `json:"sum,omitempty"`
}

type metricPoint struct {
	metricType string
	value      float64
	timestamp  time.Time
	temporality int
	monotonic   bool
	attributes  []otlpAttr
}

func extractDataPoints(m otlpMetric) []metricPoint {
	var points []metricPoint
	if m.Sum != nil {
		for _, dp := range m.Sum.DataPoints {
			points = append(points, metricPoint{
				metricType:  "sum",
				value:       dp.Value(),
				timestamp:   dp.TimeUnixNano.Time(),
				temporality: m.Sum.AggregationTemporality,
				monotonic:   m.Sum.IsMonotonic,
				attributes:  dp.Attributes,
			})
		}
	}
	if m.Gauge != nil {
		for _, dp := range m.Gauge.DataPoints {
			points = append(points, metricPoint{
				metricType: "gauge",
				value:      dp.Value(),
				timestamp:  dp.TimeUnixNano.Time(),
				attributes: dp.Attributes,
			})
		}
	}
	if m.Histogram != nil {
		for _, dp := range m.Histogram.DataPoints {
			v := 0.0
			if dp.Sum != nil {
				v = *dp.Sum
			}
			points = append(points, metricPoint{
				metricType:  "histogram",
				value:       v,
				timestamp:   dp.TimeUnixNano.Time(),
				temporality: m.Histogram.AggregationTemporality,
				attributes:  dp.Attributes,
			})
		}
	}
	return points
}

func extractServiceName(attrs []otlpAttr) string {
	for _, a := range attrs {
		if a.Key == "service.name" {
			return a.Value.String()
		}
	}
	return "unknown"
}

func flattenAttrs(attrs []otlpAttr) map[string]string {
	m := make(map[string]string, len(attrs))
	for _, a := range attrs {
		m[a.Key] = a.Value.String()
	}
	return m
}

func nullStr(s string) interface{} {
	if s == "" {
		return nil
	}
	return s
}

File: crates/lotel-storage/src/ingest.rs
use std::io::{BufRead, BufReader};
use std::path::Path;

use anyhow::{Context, Result};
use duckdb::{Connection, Transaction};
use serde::Deserialize;
use serde_json::Value;

/// Ingest all JSONL files from data_path into the database.
pub fn ingest_all(conn: &Connection, data_path: &Path) -> Result<()> {
    for (signal, ingest_fn) in [
        ("traces", ingest_traces as fn(&Connection, &Path) -> Result<()>),
        ("metrics", ingest_metrics as fn(&Connection, &Path) -> Result<()>),
        ("logs", ingest_logs as fn(&Connection, &Path) -> Result<()>),
    ] {
        let file = data_path.join(signal).join(format!("{signal}.jsonl"));
        if file.exists() {
            ingest_fn(conn, &file).with_context(|| format!("ingesting {signal}"))?;
        }
    }
    Ok(())
}

/// Nanosecond timestamp that handles both string and integer JSON representations.
#[derive(Debug, Clone, Copy, Default, Deserialize)]
#[serde(untagged)]
enum OtlpNano {
    #[default]
    Missing,
    Int(i64),
    Str(#[serde(deserialize_with = "deserialize_nano_str")] i64),
}

fn deserialize_nano_str<'de, D: serde::Deserializer<'de>>(d: D) -> Result<i64, D::Error> {
    let s = String::deserialize(d)?;
    Ok(s.parse().unwrap_or(0))
}

impl OtlpNano {
    fn to_datetime(self) -> Option<chrono::NaiveDateTime> {
        let ns = match self {
            OtlpNano::Int(n) => n,
            OtlpNano::Str(n) => n,
            OtlpNano::Missing => return None,
        };
        if ns == 0 {
            return None;
        }
        let secs = ns / 1_000_000_000;
        let nsecs = (ns % 1_000_000_000) as u32;
        chrono::DateTime::from_timestamp(secs, nsecs).map(|dt| dt.naive_utc())
    }
}

// --- OTLP JSON structures ---

#[derive(Deserialize)]
struct OtlpAttr {
    key: String,
    value: Option<OtlpValue>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct OtlpValue {
    #[serde(alias = "string_value")]
    string_value: Option<String>,
    #[serde(alias = "int_value")]
    int_value: Option<Value>,
    #[serde(alias = "bool_value")]
    bool_value: Option<bool>,
    #[serde(alias = "double_value")]
    double_value: Option<f64>,
}

impl OtlpValue {
    fn as_string(&self) -> String {
        if let Some(s) = &self.string_value {
            return s.clone();
        }
        if let Some(v) = &self.int_value {
            return match v {
                Value::String(s) => s.clone(),
                Value::Number(n) => n.to_string(),
                _ => String::new(),
            };
        }
        if let Some(b) = self.bool_value {
            return b.to_string();
        }
        if let Some(d) = self.double_value {
            return format!("{d}");
        }
        String::new()
    }
}

fn extract_service_name(attrs: &[OtlpAttr]) -> String {
    for attr in attrs {
        if attr.key == "service.name" {
            if let Some(v) = &attr.value {
                return v.as_string();
            }
        }
    }
    "unknown".to_string()
}

fn flatten_attrs(attrs: &[OtlpAttr]) -> serde_json::Value {
    let mut map = serde_json::Map::new();
    for attr in attrs {
        let val = attr
            .value
            .as_ref()
            .map(|v| v.as_string())
            .unwrap_or_default();
        map.insert(attr.key.clone(), Value::String(val));
    }
    Value::Object(map)
}

// --- Traces ingestion ---

// Note: we use rename_all="camelCase" to match standard OTLP JSON format (from Go OTel collector),
// and add alias attributes for snake_case to also accept proto serde output (from Rust exporter).

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct TraceBatch {
    #[serde(alias = "resource_spans")]
    resource_spans: Vec<ResourceSpan>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct ResourceSpan {
    resource: Option<Resource>,
    #[serde(alias = "scope_spans")]
    scope_spans: Vec<ScopeSpan>,
}

#[derive(Deserialize)]
struct Resource {
    attributes: Option<Vec<OtlpAttr>>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct ScopeSpan {
    spans: Vec<SpanJson>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct SpanJson {
    #[serde(alias = "trace_id")]
    trace_id: Option<String>,
    #[serde(alias = "span_id")]
    span_id: Option<String>,
    #[serde(alias = "parent_span_id")]
    parent_span_id: Option<String>,
    name: Option<String>,
    kind: Option<i32>,
    #[serde(default, alias = "start_time_unix_nano")]
    start_time_unix_nano: OtlpNano,
    #[serde(default, alias = "end_time_unix_nano")]
    end_time_unix_nano: OtlpNano,
    status: Option<SpanStatus>,
    attributes: Option<Vec<OtlpAttr>>,
}

#[derive(Deserialize)]
struct SpanStatus {
    code: Option<i32>,
}

fn ingest_traces(conn: &Connection, file: &Path) -> Result<()> {
    let f = std::fs::File::open(file)?;
    let reader = BufReader::with_capacity(1024 * 1024, f);

    let tx = conn.unchecked_transaction()?;

    for line in reader.lines() {
        let line = line?;
        if line.trim().is_empty() {
            continue;
        }

        let batch: TraceBatch = match serde_json::from_str(&line) {
            Ok(b) => b,
            Err(_) => continue, // Skip malformed lines.
        };

        for rs in &batch.resource_spans {
            let svc_name = rs
                .resource
                .as_ref()
                .and_then(|r| r.attributes.as_ref())
                .map(|a| extract_service_name(a))
                .unwrap_or_else(|| "unknown".to_string());

            for ss in &rs.scope_spans {
                for span in &ss.spans {
                    insert_span(&tx, span, &svc_name)?;
                }
            }
        }
    }

    tx.commit()?;
    Ok(())
}

fn insert_span(tx: &Transaction, span: &SpanJson, svc_name: &str) -> Result<()> {
    let start_time = span.start_time_unix_nano.to_datetime();
    let end_time = span.end_time_unix_nano.to_datetime();
    let duration_ns = match (start_time, end_time) {
        (Some(s), Some(e)) => (e - s).num_nanoseconds().unwrap_or(0),
        _ => 0,
    };
    let attrs = span
        .attributes
        .as_ref()
        .map(|a| flatten_attrs(a))
        .unwrap_or(Value::Object(serde_json::Map::new()));
    let attrs_str = serde_json::to_string(&attrs)?;
    let date_str = start_time.map(|t| t.format("%Y-%m-%d").to_string());

    tx.execute(
        "INSERT INTO traces (trace_id, span_id, parent_span_id, name, kind, start_time, end_time, duration_ns, status_code, service_name, attributes, date) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
        duckdb::params![
            span.trace_id.as_deref().unwrap_or(""),
            span.span_id.as_deref().unwrap_or(""),
            span.parent_span_id.as_deref(),
            span.name.as_deref().unwrap_or(""),
            span.kind.unwrap_or(0),
            start_time,
            end_time,
            duration_ns,
            span.status.as_ref().and_then(|s| s.code).unwrap_or(0),
            svc_name,
            attrs_str,
            date_str.as_deref(),
        ],
    )?;
    Ok(())
}

// --- Metrics ingestion ---

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct MetricBatch {
    #[serde(alias = "resource_metrics")]
    resource_metrics: Vec<ResourceMetric>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct ResourceMetric {
    resource: Option<Resource>,
    #[serde(alias = "scope_metrics")]
    scope_metrics: Vec<ScopeMetric>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct ScopeMetric {
    metrics: Vec<MetricJson>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct MetricJson {
    name: String,
    unit: Option<String>,
    sum: Option<SumJson>,
    gauge: Option<GaugeJson>,
    histogram: Option<HistogramJson>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct SumJson {
    #[serde(alias = "data_points")]
    data_points: Vec<DataPointJson>,
    #[serde(alias = "aggregation_temporality")]
    aggregation_temporality: Option<i32>,
    #[serde(alias = "is_monotonic")]
    is_monotonic: Option<bool>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct GaugeJson {
    #[serde(alias = "data_points")]
    data_points: Vec<DataPointJson>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct HistogramJson {
    #[serde(alias = "data_points")]
    data_points: Vec<HistogramDPJson>,
    #[serde(alias = "aggregation_temporality")]
    aggregation_temporality: Option<i32>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct DataPointJson {
    attributes: Option<Vec<OtlpAttr>>,
    #[serde(default, alias = "time_unix_nano")]
    time_unix_nano: OtlpNano,
    #[serde(alias = "as_int")]
    as_int: Option<Value>,
    #[serde(alias = "as_double")]
    as_double: Option<f64>,
}

impl DataPointJson {
    fn value(&self) -> f64 {
        if let Some(d) = self.as_double {
            return d;
        }
        if let Some(v) = &self.as_int {
            return match v {
                Value::String(s) => s.parse().unwrap_or(0.0),
                Value::Number(n) => n.as_f64().unwrap_or(0.0),
                _ => 0.0,
            };
        }
        0.0
    }
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct HistogramDPJson {
    attributes: Option<Vec<OtlpAttr>>,
    #[serde(default, alias = "time_unix_nano")]
    time_unix_nano: OtlpNano,
    sum: Option<f64>,
}

struct MetricPoint {
    metric_type: &'static str,
    value: f64,
    timestamp: Option<chrono::NaiveDateTime>,
    temporality: Option<i32>,
    monotonic: Option<bool>,
    attributes: serde_json::Value,
}

fn extract_data_points(m: &MetricJson) -> Vec<MetricPoint> {
    let mut points = Vec::new();

    if let Some(sum) = &m.sum {
        for dp in &sum.data_points {
            points.push(MetricPoint {
                metric_type: "sum",
                value: dp.value(),
                timestamp: dp.time_unix_nano.to_datetime(),
                temporality: sum.aggregation_temporality,
                monotonic: sum.is_monotonic,
                attributes: dp
                    .attributes
                    .as_ref()
                    .map(|a| flatten_attrs(a))
                    .unwrap_or(Value::Object(serde_json::Map::new())),
            });
        }
    }

    if let Some(gauge) = &m.gauge {
        for dp in &gauge.data_points {
            points.push(MetricPoint {
                metric_type: "gauge",
                value: dp.value(),
                timestamp: dp.time_unix_nano.to_datetime(),
                temporality: None,
                monotonic: None,
                attributes: dp
                    .attributes
                    .as_ref()
                    .map(|a| flatten_attrs(a))
                    .unwrap_or(Value::Object(serde_json::Map::new())),
            });
        }
    }

    if let Some(hist) = &m.histogram {
        for dp in &hist.data_points {
            points.push(MetricPoint {
                metric_type: "histogram",
                value: dp.sum.unwrap_or(0.0),
                timestamp: dp.time_unix_nano.to_datetime(),
                temporality: hist.aggregation_temporality,
                monotonic: None,
                attributes: dp
                    .attributes
                    .as_ref()
                    .map(|a| flatten_attrs(a))
                    .unwrap_or(Value::Object(serde_json::Map::new())),
            });
        }
    }

    points
}

fn ingest_metrics(conn: &Connection, file: &Path) -> Result<()> {
    let f = std::fs::File::open(file)?;
    let reader = BufReader::with_capacity(1024 * 1024, f);

    let tx = conn.unchecked_transaction()?;

    for line in reader.lines() {
        let line = line?;
        if line.trim().is_empty() {
            continue;
        }

        let batch: MetricBatch = match serde_json::from_str(&line) {
            Ok(b) => b,
            Err(_) => continue,
        };

        for rm in &batch.resource_metrics {
            let svc_name = rm
                .resource
                .as_ref()
                .and_then(|r| r.attributes.as_ref())
                .map(|a| extract_service_name(a))
                .unwrap_or_else(|| "unknown".to_string());

            for sm in &rm.scope_metrics {
                for m in &sm.metrics {
                    for dp in extract_data_points(m) {
                        let attrs_str = serde_json::to_string(&dp.attributes)?;
                        let date_str = dp.timestamp.map(|t| t.format("%Y-%m-%d").to_string());

                        tx.execute(
                            "INSERT INTO metrics (metric_name, metric_type, value, timestamp, service_name, aggregation_temporality, is_monotonic, unit, attributes, date) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                            duckdb::params![
                                m.name,
                                dp.metric_type,
                                dp.value,
                                dp.timestamp,
                                svc_name,
                                dp.temporality,
                                dp.monotonic,
                                m.unit.as_deref(),
                                attrs_str,
                                date_str.as_deref(),
                            ],
                        )?;
                    }
                }
            }
        }
    }

    tx.commit()?;
    Ok(())
}

// --- Logs ingestion ---

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct LogBatch {
    #[serde(alias = "resource_logs")]
    resource_logs: Vec<ResourceLog>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct ResourceLog {
    resource: Option<Resource>,
    #[serde(alias = "scope_logs")]
    scope_logs: Vec<ScopeLog>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct ScopeLog {
    #[serde(alias = "log_records")]
    log_records: Vec<LogRecordJson>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct LogRecordJson {
    #[serde(default, alias = "time_unix_nano")]
    time_unix_nano: OtlpNano,
    #[serde(alias = "severity_text")]
    severity_text: Option<String>,
    #[serde(alias = "severity_number")]
    severity_number: Option<i32>,
    body: Option<OtlpValue>,
    #[serde(alias = "trace_id")]
    trace_id: Option<String>,
    #[serde(alias = "span_id")]
    span_id: Option<String>,
    attributes: Option<Vec<OtlpAttr>>,
}

fn ingest_logs(conn: &Connection, file: &Path) -> Result<()> {
    let f = std::fs::File::open(file)?;
    let reader = BufReader::with_capacity(1024 * 1024, f);

    let tx = conn.unchecked_transaction()?;

    for line in reader.lines() {
        let line = line?;
        if line.trim().is_empty() {
            continue;
        }

        let batch: LogBatch = match serde_json::from_str(&line) {
            Ok(b) => b,
            Err(_) => continue,
        };

        for rl in &batch.resource_logs {
            let svc_name = rl
                .resource
                .as_ref()
                .and_then(|r| r.attributes.as_ref())
                .map(|a| extract_service_name(a))
                .unwrap_or_else(|| "unknown".to_string());

            for sl in &rl.scope_logs {
                for lr in &sl.log_records {
                    let ts = lr.time_unix_nano.to_datetime();
                    let attrs = lr
                        .attributes
                        .as_ref()
                        .map(|a| flatten_attrs(a))
                        .unwrap_or(Value::Object(serde_json::Map::new()));
                    let attrs_str = serde_json::to_string(&attrs)?;
                    let body_str = lr.body.as_ref().map(|b| b.as_string());
                    let date_str = ts.map(|t| t.format("%Y-%m-%d").to_string());
                    let trace_id = lr.trace_id.as_deref().filter(|s| !s.is_empty());
                    let span_id = lr.span_id.as_deref().filter(|s| !s.is_empty());

                    tx.execute(
                        "INSERT INTO logs (timestamp, severity, severity_number, body, service_name, trace_id, span_id, attributes, date) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
                        duckdb::params![
                            ts,
                            lr.severity_text.as_deref(),
                            lr.severity_number,
                            body_str.as_deref(),
                            svc_name,
                            trace_id,
                            span_id,
                            attrs_str,
                            date_str.as_deref(),
                        ],
                    )?;
                }
            }
        }
    }

    tx.commit()?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db;
    use std::io::Write;

    fn setup_db() -> Connection {
        db::open_in_memory().unwrap()
    }

    #[test]
    fn ingest_traces_jsonl() {
        let conn = setup_db();
        let tmp = tempfile::TempDir::new().unwrap();
        let traces_dir = tmp.path().join("traces");
        std::fs::create_dir_all(&traces_dir).unwrap();
        let file = traces_dir.join("traces.jsonl");

        let data = r#"{"resourceSpans":[{"resource":{"attributes":[{"key":"service.name","value":{"stringValue":"test-svc"}}]},"scopeSpans":[{"spans":[{"traceId":"abc123","spanId":"def456","name":"test-span","kind":1,"startTimeUnixNano":"1710000000000000000","endTimeUnixNano":"1710000001000000000","status":{"code":0},"attributes":[{"key":"http.method","value":{"stringValue":"GET"}}]}]}]}]}"#;
        std::fs::write(&file, format!("{data}\n")).unwrap();

        ingest_traces(&conn, &file).unwrap();

        let count: i64 = conn
            .query_row("SELECT COUNT(*) FROM traces", [], |row| row.get(0))
            .unwrap();
        assert_eq!(count, 1);

        let name: String = conn
            .query_row("SELECT name FROM traces LIMIT 1", [], |row| row.get(0))
            .unwrap();
        assert_eq!(name, "test-span");

        let svc: String = conn
            .query_row("SELECT service_name FROM traces LIMIT 1", [], |row| {
                row.get(0)
            })
            .unwrap();
        assert_eq!(svc, "test-svc");
    }

    #[test]
    fn ingest_metrics_jsonl() {
        let conn = setup_db();
        let tmp = tempfile::TempDir::new().unwrap();
        let metrics_dir = tmp.path().join("metrics");
        std::fs::create_dir_all(&metrics_dir).unwrap();
        let file = metrics_dir.join("metrics.jsonl");

        let data = r#"{"resourceMetrics":[{"resource":{"attributes":[{"key":"service.name","value":{"stringValue":"test-svc"}}]},"scopeMetrics":[{"metrics":[{"name":"http.requests","unit":"1","sum":{"dataPoints":[{"timeUnixNano":"1710000000000000000","asDouble":42.0,"attributes":[]}],"aggregationTemporality":2,"isMonotonic":true}}]}]}]}"#;
        std::fs::write(&file, format!("{data}\n")).unwrap();

        ingest_metrics(&conn, &file).unwrap();

        let count: i64 = conn
            .query_row("SELECT COUNT(*) FROM metrics", [], |row| row.get(0))
            .unwrap();
        assert_eq!(count, 1);

        let value: f64 = conn
            .query_row("SELECT value FROM metrics LIMIT 1", [], |row| row.get(0))
            .unwrap();
        assert!((value - 42.0).abs() < f64::EPSILON);
    }

    #[test]
    fn ingest_logs_jsonl() {
        let conn = setup_db();
        let tmp = tempfile::TempDir::new().unwrap();
        let logs_dir = tmp.path().join("logs");
        std::fs::create_dir_all(&logs_dir).unwrap();
        let file = logs_dir.join("logs.jsonl");

        let data = r#"{"resourceLogs":[{"resource":{"attributes":[{"key":"service.name","value":{"stringValue":"test-svc"}}]},"scopeLogs":[{"logRecords":[{"timeUnixNano":"1710000000000000000","severityText":"INFO","severityNumber":9,"body":{"stringValue":"hello world"},"attributes":[]}]}]}]}"#;
        std::fs::write(&file, format!("{data}\n")).unwrap();

        ingest_logs(&conn, &file).unwrap();

        let count: i64 = conn
            .query_row("SELECT COUNT(*) FROM logs", [], |row| row.get(0))
            .unwrap();
        assert_eq!(count, 1);

        let body: String = conn
            .query_row("SELECT body FROM logs LIMIT 1", [], |row| row.get(0))
            .unwrap();
        assert_eq!(body, "hello world");
    }

    #[test]
    fn ingest_all_skips_missing() {
        let conn = setup_db();
        let tmp = tempfile::TempDir::new().unwrap();
        // No files exist — should not error.
        ingest_all(&conn, tmp.path()).unwrap();
    }
}
